### ISSUE TITLE
Purge image cache when shelving

### ISSUE THREAD

**Issue by edsu** [sul-dlss/dor-services-app#4146]
There is a [long standing need](https://github.com/sul-dlss/stacks/issues/743) to allow updates to images in SDR objects to be reflected when viewing images in Stacks. Unfortunately stale images can persist on the web for days, and sometimes weeks, before the updated images become available. This behavior is problematic for Argo users who would like to confirm the success of their changes, and also for the public who do not get to see the updated content for some indeterminate amount of time. One example that came up recently is when @peterchanws has been regenerating Web Archive Seed thumbnails, which are not getting reflected in Argo or SearchWorks.

One theory (yet to be tested) is that [Cantaloupe](https://cantaloupe-project.github.io) (the software that is used to provide the image service) is holding on to tile derivatives after the source image has been changed.  Now that Cantaloupe has an API endpoint to support purging derivative tiles from its cache ([PurgeItemFromCache](https://cantaloupe-project.github.io/manual/5.0/endpoints.html#PurgeItemFromCache)) it ought to be possible to purge images when they need to be refreshed.

@mjgiarlo and @cbeer discussed where it might make sense to make this API call, and in the absence of a Shelving API, were thinking that it could make sense to try purging the image [during shelving](https://github.com/sul-dlss/dor-services-app/blob/main/app/services/shelving_service.rb#L19-L34) here in DSA.

But before we do that we would want to verify that purging from Cantaloupe actually has the desired effect and that there aren't more levels of caching within the [image serving architecture](https://consul.stanford.edu/display/digistacks/Image+Stacks) that are at play.

It appears that purging should be achievable with the equivalent of (for staging environment):

```
curl --include \
     --header "Content-Type: application/json" \
     --request POST \
     --user '{username}:{password}' \
     --data '{"verb": "PurgeItemFromCache", "identifier": "{identifier}"}' \
     https://imageserver-stage.stanford.edu/tasks
```

While `username` and `password` can be [determined](https://github.com/sul-dlss/puppet/blob/21a406ed9e629646b884d5cf42c81ad8c784c7d9/modules/profile/files/cantaloupe/cantaloupe505.properties#L119-L120) but it's not yet known what should be used for `identifier`.

**Comment by mjgiarlo** [comment]
Not exactly blocked by https://github.com/sul-dlss/operations-tasks/issues/3094 but whoever picks this up should keep an eye on that issue since it will change the URL to HTTPS and likely change the credentials.

**Comment by mjgiarlo** [comment]
The image servers now support requests via HTTPS.

**Comment by mjgiarlo** [comment]
The password for the APIs has changed and is now behind our vault instance. The key to look up to retrieve the value is `vault_imageserver/cantaloupe_http_secret`.

**Comment by edsu** [comment]
@justinlittman if you are looking for an image to test with, this item:

https://argo-stage.stanford.edu/view/druid:fg981bn7530

had a problem when its thumbnail was initially generated, but has subsequently had a new version created with a corrected thumbnail, which you can see in the updated JP2:

https://stacks-stage.stanford.edu/file/druid:fg981bn7530/thumbnail.jp2

The thumbnail in [SearchWorks](https://searchworks-preview-stage.stanford.edu/view/fg981bn7530) and Argo Stage still show the initial error image from the previous JP2. 

Does that help with testing at all?

**Comment by justinlittman** [comment]
I've tried multiple variations of:
```
curl --include \
     --header "Content-Type: application/json" \
     --request POST \
     --user 'admin:KS********' \
     --data '{"verb": "PurgeItemFromCache", "identifier": "fg981bn7530/thumbnail.jpg"}' \
     https://imageserver-stage.stanford.edu/tasks
```
with `/thumbnail`, `/thumbnail.jp2`, etc to no avail. In each case, Cantaloupe accepts the task and reports the task as a success. However, the thumbnail images does not change.

It is unclear whether the purge request is incorrect or there is additional caching elsewhere (Varnish, Rails).

I'm open to additional suggestions. Can we reach out to Access for assistance?

**Comment by edsu** [comment]
Yes, I ran into the same issue with the identifier and was hoping you might've hit on something I didn't.  I was thinking it could help to find where images were being added, to see what identifiers were being used, but I wasn't quite sure where to go looking for that.

I did noticed this [in the docs](https://cantaloupe-project.github.io/manual/5.0/caching.html#Maintenance):

> Maintenance

Because cached content is not automatically deleted after becoming invalid, there will likely be a certain amount of invalid content taking up space in the cache at any given time. Without periodic maintenance, the amount can only grow. If this is a problem, it can be dealt with manually using the [HTTP API](https://cantaloupe-project.github.io/manual/5.0/endpoints.html#HTTP%20API), or automatically using the cache worker, which periodically purges invalid items. (See the cache.server.worker.* configuration options.)

This makes me wonder if we need to enable the cache server worker to do the purging? It appears to be currently disabled?

https://github.com/sul-dlss/puppet/blob/production/modules/profile/files/cantaloupe/cantaloupe415.properties#L496

**Comment by mjgiarlo** [comment]
@edsu Interesting. My read on that is that the worker can be turned on to automate deletion of cached content beyond a certain age, but that the worker is not required if the HTTP API is used to do it "manually." I could be wrong, of course.

**Comment by edsu** [comment]
Yeah, I can definitely see that. Maybe I'm reading between the lines too much, but the HTTP response from the API call kind of makes it sound like the purge request was queued?

I was thinking it might be easier to test a dev instance of [Cantaloupe](https://github.com/cantaloupe-project/cantaloupe), add an image, purge an image, and see if it behaved normally there before extrapolating to our qa/stage/prod environments.

**Comment by edsu** [comment]
Yeah and it doesn't really seem to work? 

https://stacks-stage.stanford.edu/image/iiif/fg981bn7530%2Fthumbnail/full/!400,400/0/default.jpg

doesn't seem to reliably return the right image when force-reloading?!

**Comment by edsu** [comment]
+1 for iceboxing until we can think this through a bit more.

The reason why this is tricky to test in Stage is that the two image servers are not configured to use the same NFS share for derivatives, unlike in Production where they do. So when we purged an image it was just from one of the two. 

@jcoyne suggested that ideally the version number should be part of the Cantaloupe identifier, which would obviate the need to explicitly purge the cache when a new version is published. I think this could be provide a simpler way forward especially if it only needs to be made in sul-dlss/stacks.

**Comment by mjgiarlo** [comment]
OK, iceboxed.
